Thanks for the PR.
The CI is failing:

```
error[E0599]: no function or associated item named `from_raw_parts` found for struct `ArenaIndex` in the current scope
   --> internal\backends\testing\mcp_server.rs:889:48
    |
889 |           let index = introspection::ArenaIndex::from_raw_parts(42, 7);
    |                                                  ^^^^^^^^^^^^^^ function or associated item not found in `ArenaIndex`
    |
   ::: internal\backends\testing\introspection.rs:17:1
    |
 17 | / slotmap::new_key_type! {
 18 | |     pub(crate) struct ArenaIndex;
 19 | | }
    | |_- function or associated item `from_raw_parts` not found for this struct

```

--- a/internal/backends/testing/introspection.rs
+++ b/internal/backends/testing/introspection.rs
@@ -454,8 +454,35 @@ pub(crate) fn index_to_handle(index: ArenaIndex) -> proto::Handle {
     proto::Handle { index: ffi & 0xffff_ffff, generation: ffi >> 32 }
 }
 
-pub(crate) fn handle_to_index(handle: proto::Handle) -> ArenaIndex {
-    KeyData::from_ffi((handle.generation << 32) | (handle.index & 0xffff_ffff)).into()
+pub(crate) fn handle_to_index(handle: proto::Handle) -> Result<ArenaIndex, String> {
+    if handle.index > u64::from(u32::MAX) || handle.generation > u64::from(u32::MAX) {
+        return Err("Invalid handle".to_string());
+    }
+
+    let ffi = (handle.generation << 32) | handle.index;
+    let index: ArenaIndex = KeyData::from_ffi(ffi).into();
+
+    // Reject malformed handles instead of accepting slotmap's normalization.
+    if index.data().as_ffi() != ffi {
+        return Err("Invalid handle".to_string());
+    }
+
+    Ok(index)
+}
+
+#[test]
+fn test_handle_to_index_rejects_noncanonical_generation() {
+    assert!(handle_to_index(proto::Handle { index: 42, generation: 6 }).is_err());
+}
+
+#[test]
+fn test_handle_to_index_rejects_out_of_range_parts() {
+    assert!(
+        handle_to_index(proto::Handle { index: u64::from(u32::MAX) + 1, generation: 7 }).is_err()
+    );
+    assert!(
+        handle_to_index(proto::Handle { index: 42, generation: u64::from(u32::MAX) + 1 }).is_err()
+    );
 }
 
 #[test]

--- a/internal/backends/testing/mcp_server.rs
+++ b/internal/backends/testing/mcp_server.rs
@@ -183,8 +183,9 @@ async fn handle_tool_call(
         }
         "get_window_properties" => {
             let p: proto::RequestWindowProperties = deserialize_params(args)?;
-            let window_index =
-                handle_to_index(p.window_handle.ok_or_else(|| "missing windowHandle".to_string())?);
+            let window_index = handle_to_index(
+                p.window_handle.ok_or_else(|| "missing windowHandle".to_string())?,
+            )?;
             let response = state.window_properties(window_index)?;
             Ok(ToolResult::Json(
                 serde_json::to_value(response).map_err(|e| format!("serialize error: {e}"))?,
@@ -192,8 +193,9 @@ async fn handle_tool_call(
         }
         "find_elements_by_id" => {
             let p: proto::RequestFindElementsById = deserialize_params(args)?;
-            let window_index =
-                handle_to_index(p.window_handle.ok_or_else(|| "missing windowHandle".to_string())?);
+            let window_index = handle_to_index(
+                p.window_handle.ok_or_else(|| "missing windowHandle".to_string())?,
+            )?;
             let elements = state.find_elements_by_id(window_index, &p.elements_id)?;
             let response = proto::ElementsResponse {
                 element_handles: elements
@@ -209,7 +211,7 @@ async fn handle_tool_call(
             let p: proto::RequestElementProperties = deserialize_params(args)?;
             let element_index = handle_to_index(
                 p.element_handle.ok_or_else(|| "missing elementHandle".to_string())?,
-            );
+            )?;
             let element = state.element("get_element_properties", element_index)?;
             let response = introspection::element_properties(&element);
             Ok(ToolResult::Json(
@@ -220,7 +222,7 @@ async fn handle_tool_call(
             let p: proto::RequestQueryElementDescendants = deserialize_params(args)?;
             let element_index = handle_to_index(
                 p.element_handle.ok_or_else(|| "missing elementHandle".to_string())?,
-            );
+            )?;
             let element = state.element("query_element_descendants", element_index)?;
             let results =
                 introspection::query_element_descendants(element, p.query_stack, p.find_all)?;
@@ -241,7 +243,7 @@ async fn handle_tool_call(
             let max_elements: usize =
                 if p.max_elements == 0 { 200 } else { (p.max_elements as usize).clamp(1, 1000) };
 
-            let root_index = handle_to_index(element_handle);
+            let root_index = handle_to_index(element_handle)?;
             let root_element = state.element("get_element_tree", root_index)?;
 
             let mut elements: Vec<Value> = Vec::new();
@@ -286,8 +288,9 @@ async fn handle_tool_call(
         }
         "take_screenshot" => {
             let p: proto::RequestTakeSnapshot = deserialize_params(args)?;
-            let window_index =
-                handle_to_index(p.window_handle.ok_or_else(|| "missing windowHandle".to_string())?);
+            let window_index = handle_to_index(
+                p.window_handle.ok_or_else(|| "missing windowHandle".to_string())?,
+            )?;
             let response = state.take_snapshot_response(window_index, "image/png")?;
             let png_data = response.window_contents_as_encoded_image;
             Ok(ToolResult::Image {
@@ -299,7 +302,7 @@ async fn handle_tool_call(
             let p: proto::RequestElementClick = deserialize_params(args)?;
             let element_index = handle_to_index(
                 p.element_handle.ok_or_else(|| "missing elementHandle".to_string())?,
-            );
+            )?;
             let element = state.element("click_element", element_index)?;
             let button = proto::PointerEventButton::try_from(p.button)
                 .map_err(|_| format!("invalid button value: {}", p.button))?;
@@ -319,7 +322,7 @@ async fn handle_tool_call(
             let p: proto::RequestElementDrag = deserialize_params(args)?;
             let element_index = handle_to_index(
                 p.element_handle.ok_or_else(|| "missing elementHandle".to_string())?,
-            );
+            )?;
             let element = state.element("drag_element", element_index)?;
             let target_pos = p.target.ok_or_else(|| "missing target position".to_string())?;
             let target = i_slint_core::api::LogicalPosition::new(target_pos.x, target_pos.y);
@@ -336,7 +339,7 @@ async fn handle_tool_call(
             let p: proto::RequestInvokeElementAccessibilityAction = deserialize_params(args)?;
             let element_index = handle_to_index(
                 p.element_handle.ok_or_else(|| "missing elementHandle".to_string())?,
-            );
+            )?;
             let element = state.element("invoke_accessibility_action", element_index)?;
             let action = proto::ElementAccessibilityAction::try_from(p.action)
                 .map_err(|_| format!("invalid action value: {}", p.action))?;
@@ -350,7 +353,7 @@ async fn handle_tool_call(
             let p: proto::RequestSetElementAccessibleValue = deserialize_params(args)?;
             let element_index = handle_to_index(
                 p.element_handle.ok_or_else(|| "missing elementHandle".to_string())?,
-            );
+            )?;
             let element = state.element("set_element_value", element_index)?;
             element.set_accessible_value(p.value);
             let response = proto::SetElementAccessibleValueResponse {};
@@ -360,8 +363,9 @@ async fn handle_tool_call(
         }
         "dispatch_key_event" => {
             let p: proto::RequestDispatchKeyEvent = deserialize_params(args)?;
-            let window_index =
-                handle_to_index(p.window_handle.ok_or_else(|| "missing windowHandle".to_string())?);
+            let window_index = handle_to_index(
+                p.window_handle.ok_or_else(|| "missing windowHandle".to_string())?,
+            )?;
             let event_type = proto::KeyEventType::try_from(p.event_type)
                 .map_err(|_| format!("invalid eventType value: {}", p.event_type))?;
             let events: Vec<i_slint_core::platform::WindowEvent> = match event_type {
@@ -886,11 +890,23 @@ mod tests {
 
     #[test]
     fn test_handle_roundtrip() {
-        let index = introspection::ArenaIndex::from_raw_parts(42, 7);
+        let index = handle_to_index(proto::Handle { index: 42, generation: 7 }).unwrap();
         let handle = index_to_handle(index);
         assert_eq!(handle.index, 42);
         assert_eq!(handle.generation, 7);
-        assert_eq!(index, handle_to_index(handle));
+        assert_eq!(index, handle_to_index(handle).unwrap());
+    }
+
+    #[test]
+    fn test_mcp_rejects_noncanonical_handle() {
+        let state = make_state();
+        let resp = block_on(handle_mcp_request(
+            &state,
+            r#"{"jsonrpc":"2.0","id":6,"method":"tools/call","params":{"name":"get_window_properties","arguments":{"windowHandle":{"index":"42","generation":"6"}}}}"#,
+        ));
+        let resp = resp.unwrap();
+        assert!(resp["result"]["isError"].as_bool().unwrap_or(false));
+        assert!(resp["result"]["content"][0]["text"].as_str().unwrap().contains("Invalid handle"));
     }
 
     #[test]

--- a/internal/backends/testing/systest.rs
+++ b/internal/backends/testing/systest.rs
@@ -74,7 +74,7 @@ impl TestingClient {
                     window_handle.ok_or_else(|| {
                         "window properties request missing window handle".to_string()
                     })?,
-                ))?)
+                )?)?)
             }
             Req::RequestFindElementsById(proto::RequestFindElementsById {
                 window_handle,
@@ -83,7 +83,7 @@ impl TestingClient {
                 let elements = self.state.find_elements_by_id(
                     handle_to_index(window_handle.ok_or_else(|| {
                         "find elements by id request missing window handle".to_string()
-                    })?),
+                    })?)?,
                     &elements_id,
                 )?;
                 Resp::Elements(proto::ElementsResponse {
@@ -129,7 +129,7 @@ impl TestingClient {
                         window_handle.ok_or_else(|| {
                             "grab window request missing window handle".to_string()
                         })?,
-                    ),
+                    )?,
                     &image_mime_type,
                 )?)
             }
@@ -174,7 +174,7 @@ impl TestingClient {
                 self.state.dispatch_window_event(
                     handle_to_index(window_handle.ok_or_else(|| {
                         "window event dispatch request missing window handle".to_string()
-                    })?),
+                    })?)?,
                     convert_window_event(event.ok_or_else(|| {
                         "window event dispatch request missing event".to_string()
                     })?)?,
@@ -210,7 +210,7 @@ impl TestingClient {
     ) -> Result<ElementHandle, String> {
         let index = handle_to_index(
             element_handle.ok_or_else(|| format!("{request} missing element handle"))?,
-        );
+        )?;
         self.state.element(request, index)
     }
 

